quixstreams/rowconsumer/buffering.py

```diff
@@ -0,0 +1,403 @@
+
+        If the group has only one topic partition, it will never be paused.
+        """
+        if len(self._partition_buffers_values) == 1:
+            # Never pause a single-partition group
+            return []
+
+        tps = []
+        for buffer in self._partition_buffers_values:
+            # Pause consuming new messages for this partition
+            # when the buffer is paused, full and not idle
+            if not buffer.paused and buffer.full() and buffer.idle() == False:
+                buffer.pause()
+                tps.append((buffer.topic, buffer.partition))
+        return tps
+
+    def resume_empty(self) -> list[tuple[str, int]]:
+        """
+        Resume the empty `PartitionBuffer`s and return them as a list of tuples.
+
+        Only previously paused partitions are resumed.
+        """
+        tps = []
+        for buffer in self._partition_buffers_values:
+            # Resume consuming new messages for this partition
+            # when the buffer is paused and either idle or not full
+            if buffer.paused and (buffer.idle() != False or not buffer.full()):
+                buffer.resume()
+                tps.append((buffer.topic, buffer.partition))
+
+        return tps
+
+    def clear(self, topic: str):
+        buffer = self._partition_buffers.get(topic)
+        if buffer is not None:
+            buffer.clear()
+
+
+class InternalConsumerBuffer:
+    def __init__(self, max_partition_buffer_size: int = 10000):
+        """
+        A buffer to align messages across different topics by timestamps and consume them
+        in-order across partitions.
+
+        Under the hood, this class groups buffered messages by partition and
+        provides API to get the message with the smallest timestamp across all assigned
+        topics with the same partition number.
+
+
+        **Note**: messages are not guaranteed to be in-order 100% of the time,
+          and they can be consumed out-of-order when the producer sends new messages with a delay,
+          and they arrive to the broker later.
+
+        How it works:
+
+        - The buffer gets partitions assigned.
+        - The Consumer feeds messages to the buffer along with the high watermarks for all assigned partitions.
+        - The Consumer calls `.pop()` to get the next message to be processed.
+            If multiple partitions with the same number are assigned, the message will be popped from the partition with the smallest next timestamp, providing in-order reads.
+        - The Consumer calls `.pause_full()` and `.resume_empty()` methods to balance the reads across all partitions.
+
+        :param max_partition_buffer_size: the maximum size of the individual topic partition buffer when the buffer is considered full.
+            It is a soft limit, and it may be exceeded in some cases.
+            When individual buffer exceeds this limit, its TP can be paused to let other partitions to be consumed too.
```

Comment:
Lines a bit too long
```suggestion
        - The Consumer feeds messages to the buffer along with the high watermarks for all
            assigned partitions.
        - The Consumer calls `.pop()` to get the next message to be processed.
            If multiple partitions with the same number are assigned, the message will be
            popped from the partition with the smallest next timestamp, providing in-order reads.
        - The Consumer calls `.pause_full()` and `.resume_empty()` methods to balance the reads
            across all partitions.

        :param max_partition_buffer_size: the maximum size of the individual topic partition
            buffer when the buffer is considered full. It is a soft limit, and it may be exceeded
            in some cases. When individual buffer exceeds this limit, its TP can be paused
            to let other partitions to be consumed too.
```